### ISSUE TITLE
FileNotFoundException on Server constructor with .net48

### ISSUE THREAD

**Issue by TFTomSun** [protobuf-net/protobuf-net.Grpc#104]
Hello,

unfortunatelly I get this exception when I try to use the library grpc native from net48:
  Message: 
    System.IO.FileNotFoundException : Error loading native library. Not found in any of the possible locations: C:\git\collaboration\private.homeautomation\_out\bin\Debug\AppGates.Net.ScreenDesigner.Tests\net48\grpc_csharp_ext.x86.dll,C:\git\collaboration\private.homeautomation\_out\bin\Debug\AppGates.Net.ScreenDesigner.Tests\net48\runtimes/win/native\grpc_csharp_ext.x86.dll,C:\git\collaboration\private.homeautomation\_out\bin\Debug\AppGates.Net.ScreenDesigner.Tests\net48\../..\runtimes/win/native\grpc_csharp_ext.x86.dll
  Stack Trace: 
    UnmanagedLibrary.FirstValidLibraryPath(String[] libraryPathAlternatives)
    UnmanagedLibrary.ctor(String[] libraryPathAlternatives)
    NativeExtension.LoadUnmanagedLibrary()
    NativeExtension.LoadNativeMethods()
    NativeExtension.ctor()
    NativeExtension.Get()
    GrpcEnvironment.GrpcNativeInit()
    GrpcEnvironment.ctor()
    GrpcEnvironment.AddRef()
    Server.ctor(IEnumerable`1 options)
    Server.ctor()

**Comment by TFTomSun** [comment]
Ah, I found the problem. The copy behavior in the Grpc.Core package is not transitive. I just referenced:
```  
protobuf-net.Grpc.Native 
```` 
but then the copy behavior doesn't kick in
    
After adding the package:
```
Grpc.Core
```
it worked fine.
I think the problem can be fixed, by moving target to the buildTransitive folder and it would probably better to move it into the  protobuf-net.Grpc.Native  package instead of leaving it in the Grpc.Core package.

**Comment by mgravell** [comment]
That sounds more like a problem with the build tools (dotnet/MSBuild); it
is *meant* to be transitive. What do I need to do in order to repro this?

**Comment by TFTomSun** [comment]
create a full .net framework project and just reference protobuf-net.Grpc.Native . Build will not copy the the native assemblies to the output. Please have a look at buildTransitive vs. build folder in build packages. 
https://docs.microsoft.com/de-de/nuget/consume-packages/package-references-in-project-files

**Comment by mgravell** [comment]
Ah, right; I think I understand now. This isn't actually a problem in the build - it is a side-effect of the test project being a *library*. This is important, as transitive dependencies are only resolved at the *application* build, since it is the *application* build that effectively pins the targets rather than using "bait and switch". You can see this if you try adding (and I realize this isn't what you necessarily want for a test, but: just try it) to the csproj:

``` xml
<OutputType>Exe</OutputType>
```

You should see that the native assemblies are now added.

So basically, I don't think there's anything I can do here; this is simply a "feature" of the transitive build rules - everything is *already there*, it just doesn't trigger for transitives until application build. Since only unit tests are really impacted by this, I'm not sure we need to bend the world to fix it...

**Comment by TFTomSun** [comment]
hmm yes, its a design decision. If you want to make that copy operation explicit during the build, you could achieve that by using the buildTransitive thing. If you want to rely on the bundling of an application to get all binaries together, you are already fine with the current solution. 

I prefer to make those thing explicit. Actually, I even embed native binaries often in the .NET assembly and deploy them at runtime, so that I don't have to rely on any of these weird build copy operations.

I would still vote for making it explicit, because it took me a while to find out what is actually wrong. This might hit others, too.
